hydra-task/src/main/java/com/addthis/hydra/task/source/DataSourceQuery.java

```diff
@@ -0,0 +1,279 @@
+
+    private static final Logger log = LoggerFactory.getLogger(DataSourceQuery.class);
+
+    private final BundleFormat format = new ListBundleFormat();
+
+    /**
+     * The tree job that is to be queried. Use
+     * the literal {@code ${hydra.task.jobid}} for a job
+     * to query itself. This field is required.
+     */
+    @Nonnull private final String jobId;
+
+    /**
+     * Optionally specify the directory
+     * of the tree for the given job. Default
+     * is "data".
+     */
+    @Nonnull private final String treeDir;
+
+    /**
+     * Path for the query. This field is required.
+     */
+    @Nonnull private final String path;
+
+    /**
+     * Operations for the query. This field is optional.
+     */
+    @Nullable private final String ops;
+
+    /**
+     * Optionally specify the field names for the columns
+     * produced by the query. Otherwise the fields will
+     * be named "0", "1", "2", etc. Default is null.
+     */
+    @Nullable private final String[] fields;
+
+    /**
+     * Maximum size of the queue that stores bundles prior to their processing.
+     * Default is 128.
+     */
+    private final int buffer;
+
+    @Time(TimeUnit.MILLISECONDS)
+    private final int pollInterval;
+
+    private final int pollCountdown;
+
+    private final BlockingQueue<Bundle> queue;
+
+    private final AtomicBoolean shuttingDown = new AtomicBoolean(false);
+    private final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+    private final AtomicReference<Throwable> firstError = new AtomicReference<>();
+
+    private QueryEngineSourceSingle queryEngine;
+
+    private QueryEngineSource.Handle queryHandle;
+
+    private File tempDir;
+
+    private boolean localInputSource;
+
+    private IndexField[] fieldsIn;
+    private AutoField[] fieldsOut;
+
+    @JsonCreator
+    public DataSourceQuery(@JsonProperty(value = "jobId", required = true) String jobId,
+                           @JsonProperty(value = "treeDir", required = true) String treeDir,
+                           @JsonProperty(value = "path", required = true) String path,
+                           @JsonProperty("ops") String ops,
+                           @JsonProperty("fields") String[] fields,
+                           @JsonProperty("buffer") int buffer,
+                           @JsonProperty("pollInterval") int pollInterval,
+                           @JsonProperty("pollCountdown") int pollCountdown) {
+        this.jobId = jobId;
+        this.treeDir = treeDir;
+        this.path = path;
+        this.ops = ops;
+        this.fields = fields;
+        this.buffer = buffer;
+        this.pollInterval = pollInterval;
+        this.pollCountdown = pollCountdown;
+        this.queue = new LinkedBlockingQueue<>(this.buffer);
+    }
+
+    @Override public void init() {
+        localInputSource = Objects.equals(System.getenv("HYDRA_JOBID"), jobId);
```

Comment:
alternatives:
- have job id be optional and have null imply local source (can also rename it to "remoteJobId")
- check that the data directory is a descendant of the working directory
- make another field called "localJobId" with the appropriate default and check against that